### PR TITLE
feat: support processor properties

### PR DESCRIPTION
**Pull request by surli**
This PR intends to fix the issue https://github.com/INRIA/spoon/issues/1537

### CODE REVIEW

**Comment by surli** [comment]
@tmortagne could you have a look to https://github.com/surli/spoon-maven-plugin/blob/2a46edbe642cb86cef9c8701d1b67d178f6670e2/README.md#how-to-pass-properties-to-a-processor and tell me if the design suits you?

**Comment by tmortagne** [comment]
That was fast :)

How would it look for a property which is a list of Strings ? A Map ? I guess the real question is: what Java types are supported ?

Looks OK in general. I guess it would have been even better to refactor a bit `<processor>` to have `<name>` and `<properties>` directly in it but not super nice it term of retro compatibility.

**Comment by surli** [comment]
> I guess the real question is: what Java types are supported ?

As stated at the end of the Readme, it only supports primitive types for now.

**Comment by tmortagne** [comment]
> As stated at the end of the Readme, it only supports primitive types for now.

Indeed fair RTFM :)

So I guess I will need to define a syntax for my configuration and parse it. Important thing is that I can pass it :)

**Comment by surli** [comment]
hmmm, actually I may be wrong here: I used java Properties so I should be able to pass list directly. Let me add some tests to ensure the behaviour on this one, I'll answer properly after it.

**Comment by tmortagne** [comment]
Properties is more or less what I had in mind for my configuration so that would be nice.

**Comment by surli** [comment]
Then I upgrade the code to support list and maps: 
https://github.com/surli/spoon-maven-plugin/blob/a093c4f2b4ccbbc0bde319e83095fea5a1fa6675/README.md#how-to-pass-properties-to-a-processor

**Comment by tmortagne** [comment]
What about a Map containing list values ? (I know I'm annoying :))

**Comment by surli** [comment]
> What about a Map containing list values ? 

Currently it will led to an error. This syntax is not accepted:

```xml
<value>key=one,two,three</value>
```

But this one is: 
```xml
<value>key="one,two,three"</value>
```
And then you can split the string yourself... 

> (I know I'm annoying :))

No, that's good to challenge me as it's a new feature and I don't think about all the cases you could need.

**Comment by tmortagne** [comment]
> And then you can split the string yourself...

Yes should not be too hard for me but I tried :)

**Comment by surli** [comment]
@monperrus for me it's finished if you could have a look and merge

File: README.md [hunk 1]
```diff
@@ -79,6 +79,82 @@ In the example below, we add processor `fr.inria.gforge.spoon.processors.CountSt
...
+        </processorProperty>
+      </processorProperties>
+    </configuration>
+</plugin>
+```
+Please note that you have to specify for which processor the properties should be used with the `name` attribute.
+Primitive types are supported as well as list and maps containing string and integers. List are created by separating values with comma:
+
+```xml
+<value>one,two,three,"value containing a, comma"</value>
```
Comment (monperrus) [inline README.md]: what about using JSON for the value? client code could then use any JSON lib to handle this.

File: README.md [hunk 2]
```diff
@@ -79,6 +79,82 @@ In the example below, we add processor `fr.inria.gforge.spoon.processors.CountSt
...
+<plugin>
+    <groupId>fr.inria.gforge.spoon</groupId>
+    <artifactId>spoon-maven-plugin</artifactId>
+    <configuration>
+      <processors>
+        <processor>my.app.pkg.ProcessorWithProperty</processor>
+      </processors>
+      <processorProperties>
+        <processorProperty>
+          <name>my.app.pkg.ProcessorWithProperty</name>
```
Comment (monperrus) [inline README.md]: I would propose to use `processorName` to be more explicit and avoid confusion with the other `name` that comes after.

**Comment by monperrus** [comment]
Super cool, thanks!

what about using JSON for the value? your code, as well as client code could then use any JSON lib to handle this, and this may be an goof option for future more advanced configuration.

**Comment by surli** [comment]
> what about using JSON for the value?

Actually I though about that later, but there's different way to do it. 

1. We consider that there's is only one potential parameter in AbstractProcessor which is a JSON and then the client has to process this parameter to get all the values he wants. 
pro: the client knows exactly what he's doing and he can use all the expressivity of JSON. 
con: More work for the client and potentially less expressivity in the processor (one generic field parameter)

2. We keep the same mechanism for expressing parameters in AbstractProcessor but instead of using a XML-based properties, we use JSON-based properties
pro: we keep expressivity in processor's parameters
con: we have to parse the JSON in spoon-maven-plugin ourselves and to guess the expected types for processor parameters

**Comment by tmortagne** [comment]
Supporting only one big String property would not be very consistent with already existing Spoon Processor configuration system.

**Comment by surli** [comment]
> Supporting only one big String property would not be very consistent with already existing Spoon Processor configuration system.

Yep but as far as I know this system is new and not used anywhere so I'm not sure it's really a problem to completely change its design if needed.

**Comment by monperrus** [comment]
many thanks @tmortagne  for the suggestion and @surli  for the implementation

**Comment by tmortagne** [comment]
Great :slightly_smiling_face:

**Pull request merged**